### PR TITLE
[TEST] Regression on CDU on setProps on shallow wrapped component

### PR DESCRIPTION
**Pull request by peanutenthusiast**
For #2020.

### CODE REVIEW

**Comment by ljharb** [comment]
Thanks! I've updated the tests and split out the "no state change" and "state change" cases, and added some more assertions.

**Comment by ljharb** [comment]
I suspect this is a bug in react's shallow renderer - I've filed https://github.com/facebook/react/issues/15075 to investigate.

File: packages/enzyme-test-suite/test/ReactWrapper-spec.jsx [hunk 1]
```diff
@@ -6854,6 +6854,123 @@ describeWithDOM('mount', () => {
...
+
+      describe('setProps calls `componentDidUpdate` when `getDerivedStateFromProps` is defined', () => {
+        class DummyComp extends PureComponent {
+          constructor(...args) {
+            super(...args);
+            this.state = { state: -1 };
+          }
+
+          static getDerivedStateFromProps({ changeState, counter }, prevState) {
+            return changeState ? { state: counter * 10 } : prevState;
```
Comment (miraage) [inline packages/enzyme-test-suite/test/ReactWrapper-spec.jsx]: Shouldn't we return `null` when state does not need to be updated? https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops

Comment (miraage) [reply packages/enzyme-test-suite/test/ReactWrapper-spec.jsx]: > `getDerivedStateFromProps` is invoked right before calling the render method, both on the initial mount and on subsequent updates. It should return an object to update the state, or **null** to update nothing

Comment (ljharb) [reply packages/enzyme-test-suite/test/ReactWrapper-spec.jsx]: Yes, that's true. However, returning the previous state should result in state not being updated either. Updating to null still fails the same.

**Comment by ljharb** [comment]
I rebased this on top of #1981, which seems related, but one of the shallow tests still fails.

**Comment by peanutenthusiast** [comment]
Added `describeIfs` to ensure the tests only run for `16.3`and up

**Comment by ljharb** [comment]
Good catch :-) we still need to make the shallow test pass, though. Got any ideas for a fix?

**Comment by peanutenthusiast** [comment]
The new commit satisfies the failing test, but are there any other conditions/test cases we should cover?

**Comment by ljharb** [comment]
Probably :-) but if all the tests are passing, and it looks reasonable, let's merge it!

**Comment by peanutenthusiast** [comment]
Alrighty 👍

**Comment by dallonf** [comment]
What would it take for this fix to get released to NPM? I think I just got bitten by this bug when updating dependencies

**Comment by ljharb** [comment]
I'm waiting on a few more PRs to land before publishing the next semver-minor of enzyme itself.

**Comment by bdwain** [comment]
@ljharb are you still waiting for more versions? It would be nice to get bugfixes out ASAP since the latest release is presumably broken. I'd be happy to do a branch of 3.9.0 with just this if it would get released as 3.9.1

**Comment by ljharb** [comment]
I might be able to pull together a patch tomorrow; but there’s still a minor I’m hoping to land first.

**Comment by jdk243** [comment]
@ljharb any estimate of when this patch is going to make it to NPM? getting some test failures related to this issue after updating to 3.9.0

**Comment by joshmanders** [comment]
Is there any update on this? This is breaking our tests.

**Comment by cedosw** [comment]
Same here ... tests are failing because of this :(

**Comment by bdwain** [comment]
@ljharb seems like releases of things are going out, but not the main enzyme package itself. Any reason why there can't be a release? The latest version has been broken for 2 months while master has been fixed. Seems like it shouldn't be terrible to release a patch version of whatever is in master.

**Comment by sulemanof** [comment]
Really need this fix to be released 😢

**Comment by joshmanders** [comment]
Doesn't seem like a priority to airbnb, best to probably move on to [React Testing Library](https://testing-library.com/docs/react-testing-library/intro)

**Comment by ljharb** [comment]
@joshmanders it is very much a priority, but thanks for your quite hostile comment. If you make another one, I'll need to block you from the airbnb organization on github.

**Comment by ljharb** [comment]
@bdwain every main enzyme release potentially creates a flurry of new issues, so it is in fact quite terrible to release it before it's ready. I'm still planning to get a few fixes in before releasing v3.10.

**Comment by paulkmiller** [comment]
> @joshmanders it is very much a priority, but thanks for your quite hostile comment. If you make another one, I'll need to block you from the airbnb organization on github.

Hostile? Wow. You need to chill.

**Comment by zackify** [comment]
This is quite unfair to threaten banning  someone for stating an alternative, after waiting over  a month for a release...

**Comment by ljharb** [comment]
@paulkmiller @zackify you may be unaware of this person's history making similar comments that warrants this response, but regardless, the hostility wasn't for posting an alternative, it was for the first part.

**Comment by joshmanders** [comment]
@ljharb both of these people know me very well. I'd like for you to point out where I've made similar comments.

**Comment by ljharb** [comment]
[v3.10.0](https://github.com/airbnb/enzyme/releases/tag/enzyme%403.10.0) has now been released.

**Pull request merged**